### PR TITLE
getItemFromIri now takes an optional context as third argument

### PR DESCRIPTION
**Pull request by soyuka**
#812

ping @dunglas 

My cs-fixer doesn't give anything to fix, waiting for travis to give me informations

### CODE REVIEW

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 1]
```diff
@@ -28,35 +29,81 @@
...
+            $context = $resourceMetadata->getCollectionOperationAttribute($options['collection_operation_name'], $context, null, true);
+        } elseif (isset($options['item_operation_name'])) {
+            $context = $resourceMetadata->getItemOperationAttribute($options['item_operation_name'], $context, null, true);
+        } else {
+            $context = $resourceMetadata->getAttribute($context);
+        }
+
+        if (empty($context['groups'])) {
+            return $options;
+        } else {
```
Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: `else` statement is useless here.

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 2]
```diff
@@ -69,8 +116,12 @@ public function applyToItem(QueryBuilder $queryBuilder, QueryNameGeneratorInterf
      * @param string       $relationAlias           the previous relation alias to keep it unique
      * @param bool         $wasLeftJoin             if the relation containing the new one had a left join, we have to force the new one to left join too
      */
-    private function joinRelations(QueryBuilder $queryBuilder, string $resourceClass, array $propertyMetadataOptions = [], string $originAlias = 'o', string &$relationAlias = 'a', bool $wasLeftJoin = false)
+    private function joinRelations(QueryBuilder $queryBuilder, string $resourceClass, array $propertyMetadataOptions = [], string $originAlias = 'o', string &$relationAlias = 'a', bool $wasLeftJoin = false, $depth = 0)
     {
+        if ($depth > $this->maxJoins) {
+            throw new \RuntimeException('Too many joined relations, might be an infinite loop! Check out serialization groups!');
```
Comment (meyerbaptiste) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: `ApiPlatform\Core\Exception\RuntimeException` instead? You have to add it in the PHPDoc.

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php (hunk 1 above)
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: nice catch

File: src/Api/IriConverterInterface.php [hunk 3]
```diff
@@ -26,12 +26,13 @@
...
      * @param string $iri
      * @param bool   $fetchData
+     * @param array  $context
      *
      * @throws InvalidArgumentException
      *
      * @return object
      */
-    public function getItemFromIri(string $iri, bool $fetchData = false);
+    public function getItemFromIri(string $iri, bool $fetchData = false, array $context = []);
```
Comment (dunglas) [inline src/Api/IriConverterInterface.php]: What do you think about moving `fetchData` to a key of the context? It's really an implementation detail linked to Doctrine, it doesn't apply in most other contexts. Doing so will made the API nicer.

Comment (soyuka) [inline src/Api/IriConverterInterface.php]: I'm +1 for this ofc doing it.

Comment (soyuka) [inline src/Api/IriConverterInterface.php]: What about operationName in the `dataProvider` while we're at it?

File: tests/Bridge/Symfony/Bundle/DependencyInjection/ConfigurationTest.php [hunk 4]
```diff
@@ -70,6 +70,7 @@ public function testDefaultConfig()
             'enable_fos_user' => false,
             'enable_nelmio_api_doc' => false,
             'enable_swagger' => true,
+            'eager_loading_max_joins' => 100,
```
Comment (Simperfit) [inline tests/Bridge/Symfony/Bundle/DependencyInjection/ConfigurationTest.php]: To be sure, shoudn't that be 30 like you said in the other PR ?

**Comment by soyuka** [comment]
Should I move `operationName` to the context too [here](https://github.com/api-platform/core/pull/822/files#diff-c289b8d48b326bf2263d64149fd975d3R57)?

**Comment by dunglas** [comment]
The operation name is not implementation specific, so I think it's better to keep it as is.

Can you fix the CS please?

**Comment by soyuka** [comment]
My god why am I always messing my rebases

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 5]
```diff
@@ -68,9 +116,16 @@ public function applyToItem(QueryBuilder $queryBuilder, QueryNameGeneratorInterf
...
      * @param bool         $wasLeftJoin             if the relation containing the new one had a left join, we have to force the new one to left join too
+     * @param int          $joinsNumber             the number of joins
+     *
+     * @throws RuntimeException when the max number of joins has been reached
      */
-    private function joinRelations(QueryBuilder $queryBuilder, string $resourceClass, array $propertyMetadataOptions = [], string $originAlias = 'o', string &$relationAlias = 'a', bool $wasLeftJoin = false)
+    private function joinRelations(QueryBuilder $queryBuilder, string $resourceClass, array $propertyMetadataOptions = [], string $originAlias = 'o', string &$relationAlias = 'a', bool $wasLeftJoin = false, $joinsNumber = 0)
     {
+        if ($joinsNumber > $this->maxJoins) {
+            throw new RuntimeException('Too many joined relations, might be an infinite loop! Check out serialization groups!');
```
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: My suggestion:

Total number of joined relations has exceeded the specified maximum. Raise the limit if necessary.

File: src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php [hunk 6]
```diff
@@ -42,6 +42,7 @@ public function getConfigTreeBuilder()
                 ->scalarNode('version')->defaultValue('0.0.0')->info('The version of the API.')->end()
                 ->scalarNode('default_operation_path_resolver')->defaultValue('api_platform.operation_path_resolver.underscore')->info('Specify the default operation path resolver to use for generating resources operations path.')->end()
                 ->scalarNode('name_converter')->defaultNull()->info('Specify a name converter to use.')->end()
+                ->scalarNode('eager_loading_max_joins')->defaultValue(30)->info('Max number of joined relations before EagerLoading throws a RuntimeException')->end()
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Bundle/DependencyInjection/Configuration.php]: I suggest:

```yaml
eager_loading: true
```

or

```yaml
eager_loading:
    # enabled: true
    max_joins: 30
```

File: src/DataProvider/ItemDataProviderInterface.php [hunk 7]
```diff
@@ -26,11 +26,12 @@
      * @param string      $resourceClass
      * @param string|null $operationName
      * @param int|string  $id
-     * @param bool        $fetchData
+     * @param array       $context       Context can contain serialization context
```
Comment (teohhanhui) [inline src/DataProvider/ItemDataProviderInterface.php]: This comment should not be in the interface. It's an implementation detail.

Comment (soyuka) [inline src/DataProvider/ItemDataProviderInterface.php]: Okay, but then it might be forgotten by future implementations.

Comment (teohhanhui) [inline src/DataProvider/ItemDataProviderInterface.php]: The comment about serialization groups should go into EagerLoadingExtension.

The comment about `fetch_data` should go into the implementation of the data provider.

Comment (soyuka) [inline src/DataProvider/ItemDataProviderInterface.php]: Done thanks!

**Comment by soyuka** [comment]
@Simperfit note that I added coverage for the 0.003 % \o/

**Comment by dunglas** [comment]
Thanks @soyuka!

File: src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php [hunk 8]
```diff
@@ -28,35 +30,94 @@
...
+
+        return ['serializer_groups' => $context['groups']];
     }
 
     /**
      * {@inheritdoc}
      */
     public function applyToCollection(QueryBuilder $queryBuilder, QueryNameGeneratorInterface $queryNameGenerator, string $resourceClass, string $operationName = null)
     {
+        if (false === $this->enabled) {
```
Comment (teohhanhui) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: Actually, this extension should be removed from the service container if it's not enabled... :smile:

Comment (soyuka) [inline src/Bridge/Doctrine/Orm/Extension/EagerLoadingExtension.php]: Okay I'm changing this asap

**Pull request merged**